### PR TITLE
bpo-32028: Fix custom print suggestion having leading whitespace in print statement

### PR DESCRIPTION
**Pull request by CuriousLearner**
This fixes the newly added print suggestion for cases when there is leading whitespace in the initial data.

I've also added a test case for this. @ncoghlan Can you please check this?

https://bugs.python.org/issue32028

### CODE REVIEW

File: Objects/exceptions.c
Comment (ncoghlan) [inline Objects/exceptions.c]: Rather than adding a new call to `_PyUnicode_XStrip`, just reorder the existing operations:

1. First strip any surrounding ASCII whitespace (i.e. do this first, instead of last)
2. Then get the length of the result (rather than the length of the original text)
3. Then skip over `PRINT_OFFSET` characters at the beginning

Comment (ncoghlan) [reply Objects/exceptions.c]: A minor point I missed when reviewing the original PR: defining and using `const int STRIP_BOTH = 2;` will make the `_PyUnicode_XStrip` call more self explanatory.

So we may as well make that change now, since the code is being modified anyway.

File: Lib/test/test_print.py
Comment (ncoghlan) [inline Lib/test/test_print.py]: Since we're stripping all leading whitespace, you may as well indent this 4 spaces relative to the target variable name (so 12 leading spaces total)

**Review by ncoghlan — CHANGES_REQUESTED** [review]
The new test case is good, and the implementation changes are headed in the right general direction. However, it should be sufficient to reorder the existing operations, rather than calling `_PyUnicode_XStrip` twice.

File: Objects/exceptions.c
Comment (CuriousLearner) [reply Objects/exceptions.c]: +1 for the `const int STRIP_BOTH = 2;`. I was explaining my patch to a few folks and I had to consult my blog for that particular thing. Few months down the line I didn't even remember that.

So, yes that would be great.

Comment (CuriousLearner) [reply Objects/exceptions.c]: @ncoghlan So, while I modified the code, one of the test cases fails with re-ordering the code. That test case is:

`test_string_with_excessive_whitespace`. Since we're only stripping the the data at the beginning, and then directly trying to extract the sub-string.

I think we may need to strip the leading chars to make the previous test case pass. What do you say?

cc @serhiy-storchaka

Comment (CuriousLearner) [reply Objects/exceptions.c]: I've pushed my current code which re-orders the existing operations & the `test_string_with_excessive_whitespace` is failing. I'm waiting for your reply on this. I think we might want to strip the leading whitespace from `data`.

File: Objects/exceptions.c
Comment (serhiy-storchaka) [inline Objects/exceptions.c]: initial_data can be NULL.

File: Objects/exceptions.c
Comment (serhiy-storchaka) [inline Objects/exceptions.c]: The result of PyUnicode_Substring() can be NULL.
Otherwise it will be leaked.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
Don't forget to check for errors and count references.

**Comment by CuriousLearner** [comment]
@serhiy-storchaka Nice catch! Sorry, I missed those memory leaks. I've tried to address those issues. Can you please take a pass.

**Review by serhiy-storchaka — COMMENTED** [review]
Please remove redundant empty lines.

**Comment by CuriousLearner** [comment]
@serhiy-storchaka Fixed :)

**Review by serhiy-storchaka — APPROVED** [review]

**Review by ncoghlan — APPROVED** [review]
I'd missed that the second `_PyUnicode_XStrip` call handled excess whitespace between `print` and the expression being printed, so I was wrong about that being redundant (and the regression test suite was right).

I've adjusted the NEWS entry wording, so +1 from me (I'll merge once CI finishes).

**Comment by CuriousLearner** [comment]
Thank you so much @ncoghlan & @serhiy-storchaka :)

**Comment by CuriousLearner** [comment]
@serhiy-storchaka @ncoghlan I guess we can merge this now :)

**Comment by CuriousLearner** [comment]
Hi @serhiy-storchaka @ncoghlan 

Do we need something else here? Please let me know and I'll do that :)

**Comment by ncoghlan** [comment]
Just closing & reopending to restart the CI (Appveyor's a required check now, and it didn't run properly)

**Pull request merged**
